lib/modules/manager/tox/extract.ts

```diff
@@ -0,0 +1,136 @@
+// based on https://www.python.org/dev/peps/pep-0508/#names
+import { RANGE_PATTERN } from '@renovatebot/pep440';
+import is from '@sindresorhus/is';
+import { logger } from '../../../logger';
+import { newlineRegex, regEx } from '../../../util/regex';
+import { PypiDatasource } from '../../datasource/pypi';
+import type { PackageDependency, PackageFile, Result } from '../types';
+
+function getSectionName(str: string): string {
+  const [, sectionName] = regEx(/^\s*\[\s*(\S*)\s*]\s*$/).exec(str) ?? [];
+  return sectionName;
+}
+
+/**
+ * Get the name of the parameter
+ * Do not allow starting with spaces, this is how deps arrays are declared,
+ * by indent
+ */
+function getParamName(str: string): string {
+  const param = regEx(/^([\w.\-_]+)\s*=\s*.*?\s*$/);
+  const [, parmName] = param.exec(str) ?? [];
+  return parmName;
+}
+
+/**
+ * Taken from the setup.cfg manager due to some edits to the dep type.
+ */
+function parseDep(line: string): PackageDependency | null {
+  const packagePattern = '[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9._-]*[a-zA-Z0-9]';
+  const extrasPattern = '(?:\\s*\\[[^\\]]+\\])?';
+
+  const rangePattern: string = RANGE_PATTERN;
+  const specifierPartPattern = `\\s*${rangePattern.replace(
+    regEx(/\?<\w+>/g),
+    '?:'
+  )}`;
+  const specifierPattern = `${specifierPartPattern}(?:\\s*,${specifierPartPattern})*`;
+  const dependencyPattern = `(${packagePattern})(${extrasPattern})(${specifierPattern})`;
+
+  const pkgRegex = regEx(`^(${packagePattern})$`);
+  const pkgValRegex = regEx(`^${dependencyPattern}$`);
+
+  const [lineNoEnvMarkers] = line.split(';').map((part) => part.trim());
+  const packageMatches =
+    pkgValRegex.exec(lineNoEnvMarkers) ?? pkgRegex.exec(lineNoEnvMarkers);
+
+  if (!packageMatches) {
+    return null;
+  }
+
+  const [, depName, , currVal] = packageMatches;
+  const currentValue = currVal?.trim();
+
+  const dep: PackageDependency = {
+    depName,
+    currentValue,
+    datasource: PypiDatasource.id,
+    depType: 'test',
+  };
+
+  if (currentValue?.startsWith('==')) {
+    dep.currentVersion = currentValue.replace(/^==\s*/, '');
+  }
+
+  return dep;
+}
+
+/**
+ * Map over the lines keeping track if the section is a test env
+ * and the parameter is deps =
+ *
+ * It is only valid to have a single dep with
+ *
+ * deps = <dep>
+ *
+ * or deps with indents like so
+ *
+ * deps =
+ *     dep1
+ *     dep2
+ *
+ * Attempting to use use npm ini does not parse deps arrays correctly.
+ */
+export function extractPackageFile(
+  content: string
+): Result<PackageFile | null> {
+  logger.trace('tox.extractPackageFile()');
+
+  let sectionName: string | null = null;
+  let paramName: string | null = null;
+  let inDeps = false;
+  let isTestEnv = false;
+
+  const deps: PackageDependency[] = [];
+  const testEnvRegex = regEx(/^(testenv)\S*\s*$/);
+
+  content
```

Comment:
do you think we should use a built in parser for the given format of `tox.ini`?
according to https://tox.wiki/en/latest/config.html
there's a config parser for ini files, the official one is in python given in the site with a [link](https://docs.python.org/3/library/configparser.html)

There are Parsers for almost  every format out there
just at first glance i already see this https://www.npmjs.com/package/config-ini-parser

this could save you lots of code, bugs and flags like `inTest` and `inDeps`

Reply:
we already have an ini parser in our deps

Reply:
I agree... so gave them all a shot. Haven't found a viable candidate.

Already tried the ini parser from the deps but it interprets everything under the deps section as params so converts them to keys. Putting them back into proper deps would be a bit convoluted from this format. You'd never know when to stop iterating over the items and `=` back. And it could be a key.

ini

```javascript
 'testenv:basic': [Object: null prototype] {
        description: 'The normal case',
        passenv: '',
        '*': true,
        basepython: 'python3.10',
        deps: '',
        'basic1>': '3.1.24',
        'basic2>': '21.3',
        'basic3<': '21.3',
        'basic4<1.0.0': true,
        'basic5>1.0.0': true,
        basic6: '=1.0.0',
        'basic7~': '1.0.0',
        commands: '',
        'python {toxinidir}/tasks/release.py --version {posargs}': true
      },
      'testenv:single-line-deps': [Object: null prototype] {
        deps: 'single-line>=2.9.3',
        commands: '',
        'pytest \\': true,
        '--cov "{envsitepackagesdir}/tox" \\': true,
        '--cov-config "{toxinidir}/tox.ini" \\': true,
        '--junitxml {toxworkdir}/junit.{envname}.xml \\': true,
        '{posargs:.}': true
      },
```
Config ini's output seems less useful since it strips the range info. Don't know how we'd get it back.

config ini
```
      'testenv:basic': {
        description: 'The normal case',
        passenv: '',
        basepython: 'python3.10',
        deps: '',
        basic1: '3.1.24',
        basic2: 21.3,
        basic3: 21.3,
        basic: '=1.0.0',
        basic7: '1.0.0',
        commands: ''
      },
```

Fiddled around with config-ini-parser a bit and can't even get it to parse the file.

Tried https://www.npmjs.com/package/configparser and that produces a bunch of parsing errors.

Python's configparser parses is "correctly" for this use case. Respecting the indents.

```
 'testenv:basic': {'basepython': 'python3.10',
                   'commands': '\n'
                               'python {toxinidir}/tasks/release.py --version '
                               '{posargs}',
                   'deps': '\n'
                           'basic1>=3.1.24\n'
                           'basic2>=21.3\n'
                           'basic3<=21.3\n'
                           'basic4<1.0.0\n'
                           'basic5>1.0.0\n'
                           'basic6==1.0.0\n'
                           'basic7~=1.0.0',
                   'description': 'The normal case',
                   'passenv': '\n*'},
 'testenv:single-line-deps': {'commands': '\n'
                                          'pytest \\\n'
                                          '--cov "{envsitepackagesdir}/tox" '
                                          '\\\n'
                                          '--cov-config "{toxinidir}/tox.ini" '
                                          '\\\n'
                                          '--junitxml '
                                          '{toxworkdir}/junit.{envname}.xml '
                                          '\\\n'
                                          '{posargs:.}',
                              'deps': 'single-line>=2.9.3'},
```